Automatically create default resources in the editor for many objects

Calinou opened the pull request:
This is already done in Path2D and Path3D for the `curve` property. This applies the same principle to a lot more nodes so that you don't need as many clicks to set up a new node. If you don't need the default-created resource, load another resource in its place (the number of clicks is the same as before).

The resources automatically created are all resources you *must* create at some point for the node/resource to work correctly. Optional resources (e.g. those that act as overrides) are not created automatically to avoid any undesired behavior or performance impact.

This PR also allows `PROPERTY_USAGE_EDITOR_INSTANTIATE_OBJECT` to work with resource hints that list multiple classes `Like,This`. The first class is always selected in this case, as it's assumed to be the most commonly needed one (and is also the first one to display in the dropdown).

This behavior is editor-only, so it doesn't affect nodes/resources created from a script.

For GPUParticles3D, we can benefit further from having custom code to automatically create a billboarded QuadMesh on editor instantiation in a future PR. This code works well when placed at the end of its constructor, but it should be called by something like [`EditorData::instantiate_object_properties()`](https://github.com/godotengine/godot/blob/16d61427cab3a8e43f0a9a8ee724fc176b6433c6/editor/editor_data.cpp#L586-L599) instead so it's not considered as a default value:

```cpp
#ifdef TOOLS_ENABLED
#include "scene/resources/material.h"
#include "scene/resources/primitive_meshes.h"
#endif

// ...

#ifdef TOOLS_ENABLED
	if (Engine::get_singleton()->is_editor_hint() && get_draw_pass_mesh(0) == nullptr) {
		Ref<QuadMesh> quad_mesh = memnew(QuadMesh);
		quad_mesh->set_size(Vector2(0.05, 0.05));

		// Create material to display billboarded texture that follows the color and scale ramps
		// defined in ParticleProcessMaterial.
		Ref<StandardMaterial3D> material = memnew(StandardMaterial3D);
		material->set_shading_mode(BaseMaterial3D::SHADING_MODE_UNSHADED);
		material->set_transparency(BaseMaterial3D::TRANSPARENCY_ALPHA);
		material->set_flag(BaseMaterial3D::FLAG_ALBEDO_FROM_VERTEX_COLOR, true);
		material->set_flag(BaseMaterial3D::FLAG_BILLBOARD_KEEP_SCALE, true);
		material->set_billboard_mode(BaseMaterial3D::BILLBOARD_PARTICLES);

		// Assign a smooth circle texture.
		Ref<GradientTexture2D> texture = memnew(GradientTexture2D);
		texture->set_fill(GradientTexture2D::FILL_RADIAL);
		texture->set_fill_from(Vector2(0.5, 0.5));
		texture->set_fill_to(Vector2(0.5, 0.01));
		Ref<Gradient> gradient = memnew(Gradient);
		gradient->set_color(0, Color(1, 1, 1));
		gradient->set_color(1, Color(1, 1, 1, 0));
		texture->set_gradient(gradient);
		material->set_texture(BaseMaterial3D::TEXTURE_ALBEDO, texture);
		quad_mesh->set_material(material);

		set_draw_pass_mesh(0, quad_mesh);
	}
#endif
```

**Edit:** The above part can be implemented using https://github.com/godotengine/godot/pull/122058.

Calinou committed:
Automatically create default resources in the editor for many objects

This is already done in Path2D and Path3D for the `curve` property.
This applies the same principle to a lot more nodes so that you
don't need as many clicks to set up a new node. If you don't need
the default-created resource, load another resource in its place
(the number of clicks is the same as before).

The resources automatically created are all resources you *must* create
at some point for the node/resource to work correctly. Optional resources
(e.g. those that act as overrides) are not created automatically
to avoid any undesired behavior or performance impact.

This behavior is editor-only, so it doesn't affect nodes/resources created
from a script.

KoBeWi commented on scene/2d/tile_map_layer_group.cpp:
```diff
@@ -110,7 +110,7 @@ void TileMapLayerGroup::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_tileset"), &TileMapLayerGroup::get_tileset);
 
-	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "tile_set", PROPERTY_HINT_RESOURCE_TYPE, "TileSet"), "set_tileset", "get_tileset");
+	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "tile_set", PROPERTY_HINT_RESOURCE_TYPE, "TileSet", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_EDITOR_INSTANTIATE_OBJECT), "set_tileset", "get_tileset");
```
See #80038
TileMaps are usually used with a TileSet saved to file and shared between multiple scenes. The automatically created one is built-in, which makes it specific to the scene and also risks bloating the scene. By creating TileSet by default you are pointing users towards wrong usage of the node. It's like making Sprite2D with default empty texture.

This might apply to other resources, so you might want to re-check them with that in mind. It comes down to how the node is usually used.

Calinou replied:
I've undone the change for TileMap and GridMap.

Cammymoop commented:
Should nodes created from scripts (drag -> ctr + drop into scene tree) not get these?
It looks like scene_tree_dock.cpp doesn't call `instantiate_object_properties()` existing bug?

Cammymoop commented:
Confirmed that Path3D and Path2D are created without their default Curve when a script inheriting one of them is used to create a node by dragging and ctr + drop into the scene tree

RedMser commented on scene/resources/sky.cpp:
```diff
@@ -82,7 +82,7 @@ void Sky::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_material"), &Sky::get_material);
 
-	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "sky_material", PROPERTY_HINT_RESOURCE_TYPE, "PanoramaSkyMaterial,ProceduralSkyMaterial,PhysicalSkyMaterial,ShaderMaterial"), "set_material", "get_material");
+	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "sky_material", PROPERTY_HINT_RESOURCE_TYPE, "PanoramaSkyMaterial,ProceduralSkyMaterial,PhysicalSkyMaterial,ShaderMaterial", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_EDITOR_INSTANTIATE_OBJECT), "set_material", "get_material");
```
I'd argue procedural skies are about as common as using an environment texture, and users might not know both are even possible if it's instantiated by default. Maybe undo this particular change?

Calinou replied:
I could reorder the hint so that a ProceduralSkyMaterial is created by default. In terms of usability, I'm not fond of the fact that you need to create a Sky resource *and* a SkyMaterial resource within for a sky to be visible – but at the same time, we can't automatically create a Sky in Environment because you won't need it if you're using a plain color background (which is typical for 2D projects).

smix8 commented on scene/2d/navigation_region_2d.cpp:
```diff
@@ -336,7 +336,7 @@ void NavigationRegion2D::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("_navigation_polygon_changed"), &NavigationRegion2D::_navigation_polygon_changed);
 
-	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "navigation_polygon", PROPERTY_HINT_RESOURCE_TYPE, "NavigationPolygon"), "set_navigation_polygon", "get_navigation_polygon");
+	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "navigation_polygon", PROPERTY_HINT_RESOURCE_TYPE, "NavigationPolygon", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_EDITOR_INSTANTIATE_OBJECT), "set_navigation_polygon", "get_navigation_polygon");
```
The editor already has a dedicated popup dialog for the navigation region to add a resource when users interact with the polygon tools or baking. Auto-adding is imo not good because new users will easily have embedded resources in their scene files that can end up really big. The popup or manual add at least makes them aware of the resource.

AThousandShips commented on editor/editor_data.cpp:
```diff
@@ -592,7 +592,9 @@ void EditorData::instantiate_object_properties(Object *p_object) {
-			Object *prop = ClassDB::instantiate(pi.class_name);
+			// For comma-separated lists, instantiate the first item in the list (which is assumed to cover the most common use case).
+			// This is a situation where listing a subclass before a parent class makes sense (e.g. "NoiseTexture2D,Texture2D").
+			Object *prop = ClassDB::instantiate(String(pi.class_name).split(",")[0]);
```
```suggestion
			Object *prop = ClassDB::instantiate(String(pi.class_name).get_slicec(',', 0));
```
Should be equivalent and faster

Calinou commented:
> Confirmed that Path3D and Path2D are created without their default Curve when a script inheriting one of them is used to create a node by dragging and ctr + drop into the scene tree

This is an unrelated bug that should be tracked in its own issue.

Mickeon commented:
I think this may be worth evaluating for 4.4. The AnimatedSprites' `sprite_frames` being filled in would be particularly handy for starting developers, but it has you question to what extent the resource should be automatically created.

I am to assume it's easy to think about when there's only one option to choose.

migueldeicaza commented:
I love this change, some of the ones that we had in mind:

- CollisionShape2D should default to a shape
- MeshInstance3D should auto-pick a Mesh, and this should auto-pick a Material.

fire reviewed (approved):
I approve of automatically making objects in Godot Engine have default resources constructed as part of the usability team.

However, I have not tested every node with this new feature enabled.

Edited:

I also don't think we need to do all the nodes before merging, and we can do the additions piecemeal.

Mickeon commented:
> - CollisionShape2D should default to a shape
> - MeshInstance3D should auto-pick a Mesh, and this should auto-pick a Material.

Again, this goes back to:
> It it has you question to what extent the resource should be automatically created.

Because _always_ defaulting to the most generic shape/mesh is generally not useful, and what qualifies as "generic" is subjective.

fire commented:
For example, we can copy what Blender does. A Blender default scene has a cube, light and camera. So, focusing only on nodes, we would have the following:

`CollisionShape2D` defaults to a shape like a 2d square, and `MeshInstance3D` auto-picks a `Mesh` like a cube, which should auto-pick a `Material` like a `StandardMaterial3D`.

smix8 reviewed (approved):
I think we should merge this as is covering only the few current selected nodes where it makes a lot of sense to have a default, e.g. because they only use a single specific resource type with one option and are entirely dysfunctional without this resource added.

More objects can imo be covered by individual smaller PRs later case by case where it makes sense.

akien-mga commented:
I'm slightly concerned by generalizing this practice more (though IIRC I introduced the feature in the first place :P), as users might be unaware of this "magic" and end up having unexpected resources in their scenes.

For example if they prepare a node structure in the editor, but assign relevant resources from scripts, they'd have a dummy resource saved in the scene even though it's never going to be used. Possibly repeated a lot of times in cases like particle nodes, if they have a lot of particle emitters, etc.

smix8 commented:
Some classes already throw a popup at the user when the resource is missing and ask if the resource should be created and why (because the class does not function without). Maybe this should be added as a one-time popup for those classes so users are aware of the behavior and turn it into an editor setting so it can be changed later, aka automated for those that want and manual for those that dont.

migueldeicaza commented:
If the scenario is purely creating empty resources for the scenarios where a node needs it to even work, I dont think this will be too cumbersome or create a lot of problems.   Folks will either use it right away, or will replace it with something they actually have.

fire commented:
Remember to make it so blocking modals don't hang cicd though.

Zireael07 commented:
Does this include default Theme resources? I'm thinking something like this would help with discovering the theming system

Calinou commented:
> Does this include default Theme resources? I'm thinking something like this would help with discovering the theming system

It doesn't automatically create a Theme resource on every Control, as this would be too heavy. Only the root Control node in each scene needs to have its Theme property set for correct display.

passivestar reviewed (approved):
UX makes sense, I spent a lot of time clicking buttons when creating curves and noises.

I also don't see any issue with generalizing it to things like creating a default cube when MeshInstance3D or a CollisionShape3D are created, it's close in spirit to https://github.com/godotengine/godot/pull/119020

Calinou commented:
Rebased and tested again, it works as expected.

> I also don't see any issue with generalizing it to things like creating a default cube when MeshInstance3D or a CollisionShape3D are created, it's close in spirit to https://github.com/godotengine/godot/pull/119020

I implemented this in the latest revision of this PR, with BoxMesh and BoxShape3D respectively (RectangleShape2D in CollisionShape2D).

This also has the side effect of reordering the list of shapes in the inspector (since the first one listed is that one that's automatically instanced). This arguably makes for better UX, since rarely-used options like ArrayMesh and ImmediateMesh are now at the bottom:

### CollisionShape2D

<details>

Before | After
-|-
<img width="247" height="308" alt="Screenshot_20260623_193243" src="https://github.com/user-attachments/assets/6fbc32a7-84e8-4764-a4b2-f8c1f27a475b" /> | <img width="244" height="321" alt="Screenshot_20260623_195909" src="https://github.com/user-attachments/assets/e0866a87-47cf-4cf5-b4e1-2d19076d4c39" />
</details>

### CollisionShape3D

<details>

Before | After
-|-
<img width="244" height="346" alt="Screenshot_20260623_193258" src="https://github.com/user-attachments/assets/933db00e-45ed-4342-9c40-edd0dd552e7a" /> | <img width="244" height="346" alt="Screenshot_20260623_195337" src="https://github.com/user-attachments/assets/32c3b037-b56e-4fa8-b122-0133c3d446f1" />
</details>

### MeshInstance3D

<details>

Before | After
-|-
<img width="199" height="474" alt="Screenshot_20260623_193251" src="https://github.com/user-attachments/assets/0352283a-466a-4c9b-917b-9c7186d7e651" /> | <img width="187" height="482" alt="Screenshot_20260623_195915" src="https://github.com/user-attachments/assets/be75e289-3c82-4d8a-a9c4-46721b1c322b" />
</details>

CPUParticles3D and GPUParticles3D also automatically create a QuadMesh (since you generally use billboarded quads with them), so the list in the inspector shows QuadMesh first with BoxMesh moved to the second place.

If we want to remove the requirement to have the automatically-listed type be created first, we could add support for specifying the type to be created automatically by prefixing it with `*` in the property hint (e.g. `CircleShape2D,*RectangleShape2D`). The list order could still be used as a fallback if no type uses a `*` prefix.

KoBeWi commented:
Errors when creating NoiseTexture:
```
ERROR: Class 'Noise' or its base class cannot be instantiated.
```

KoBeWi commented on scene/3d/fog_volume.cpp:
```diff
@@ -48,7 +48,7 @@ void FogVolume::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::VECTOR3, "size", PROPERTY_HINT_RANGE, "0.01,1024,0.01,or_greater,suffix:m"), "set_size", "get_size");
 	ADD_PROPERTY(PropertyInfo(Variant::INT, "shape", PROPERTY_HINT_ENUM, "Ellipsoid (Local),Cone (Local),Cylinder (Local),Box (Local),World (Global)"), "set_shape", "get_shape");
-	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "material", PROPERTY_HINT_RESOURCE_TYPE, "FogMaterial,ShaderMaterial"), "set_material", "get_material");
+	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "material", PROPERTY_HINT_RESOURCE_TYPE, "FogMaterial,ShaderMaterial", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_EDITOR_INSTANTIATE_OBJECT), "set_material", "get_material");
```
FogVolume seems to be working fine without material (although it can't be customized). Is adding a material here something you always want to do?

Calinou replied:
> Is adding a material here something you always want to do?

In practice, yes. Otherwise, the only thing a material-less FogVolume does is show some gray fog with non-adjustable density.

KoBeWi commented on scene/3d/sprite_3d.cpp:
```diff
@@ -1552,7 +1552,7 @@ void AnimatedSprite3D::_bind_methods() {
 	ADD_SIGNAL(MethodInfo("animation_finished"));
 
-	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "sprite_frames", PROPERTY_HINT_RESOURCE_TYPE, SpriteFrames::get_class_static()), "set_sprite_frames", "get_sprite_frames");
+	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "sprite_frames", PROPERTY_HINT_RESOURCE_TYPE, SpriteFrames::get_class_static(), PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_EDITOR_INSTANTIATE_OBJECT), "set_sprite_frames", "get_sprite_frames");
```
Why AnimatedSprite2D does not get SpriteFrames?

Calinou replied:
It's an oversight, I fixed it.

KoBeWi commented on scene/resources/material.cpp:
```diff
@@ -494,7 +494,7 @@ void ShaderMaterial::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_shader_parameter", "param"), &ShaderMaterial::get_shader_parameter);
 
-	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "shader", PROPERTY_HINT_RESOURCE_TYPE, Shader::get_class_static()), "set_shader", "get_shader");
+	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "shader", PROPERTY_HINT_RESOURCE_TYPE, Shader::get_class_static(), PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_EDITOR_INSTANTIATE_OBJECT), "set_shader", "get_shader");
```
This creates an empty shader; it's not very friendly without any shader template.
Also it causes errors.

Calinou replied:
Removed it. It may not be needed anymore since we now have https://github.com/godotengine/godot/pull/99725.

KoBeWi commented:
The Noise error is still there.

Calinou commented:
> The Noise error is still there.

Fixed.

Repiteo commented:
Thanks!
